### PR TITLE
Teensy-HAL

### PR DESCRIPTION
**Pull request by teemuatlut**
New PR to the updated HAL.

Add Teensy3.5 and Teensy3.6 to the Marlin HAL.

https://travis-ci.org/teemuatlut/Marlin/branches Teensy-HAL-rebased
Travis is updated to run almost all of the checks on Teensy as well. I had to install a lot of utilities and libraries to get it all to compile automated. The Teensyduino from PJRC couldn't be used because of the GUI installer. The whole Travis check runs for about 14mins and should compile fine with both AVR and Teensy.

I've included some simple fixes for other platforms as well. The marlin serial had been moved to AVR HAL and the path needed an update. The Print Counter I took from the current RCBugFix.

I've made a new servo library that is inherited from the core, and just adds the few required additional methods. I saw no need to copy the whole library just to add a little something.

Added a waiting check for Marlin serial since the Teensy needed the while(!Serial) check to print properly on startup.

There was some really weird behavior that I saw with the commit that allowed UART and stepper ISRs inside the temperature ISR. The Teensy would completely Freeze and Windows wouldn't properly recognize it again. It also froze whatever serial communication program I was using at the time. I don't know why it was doing this but maybe someone else can look into it as well. I've since disabled sei and cli commands in the temp isr.

I'll try to add comments around to explain changes but I'd like to see others review and possibly even try it out, but I think this might be close to a good starting point.

### CODE REVIEW

File: Marlin/Marlin_main.cpp
Comment (bobc) [inline Marlin/Marlin_main.cpp]: The fundamental idea of the HAL is that all platform specific code is confined to the relevant HAL_xxx folder. The main code should not contain any references such as __MK64FX512__. Putting in platform specific references for every platform will make the code unmaintainable, and a lot harder to port to new targets.

This comment applies in several other places stepper.cpp, temperature.cpp

**Comment by bobc** [comment]
Overall, I would say this looks really good, nice work.

We need to be very careful about the use of "#if defined (\_\_MK64FX512__) ..." outside of the HAL itself.

**Review by bobc — CHANGES_REQUESTED** [review]
I suggest that we introduce a new type "HAL_Servo" or similiar. Then in servo.h, 
  `#define HAL_Servo Servo`
or
`  #define HAL_Servo libServo`
as appropriate

or perhaps better to rename libServo to HAL_Servo to avoid extra #define.

**Comment by teemuatlut** [comment]
Yes, I'd rather not see any of the platform specific compiler directives within the code. But this would've required me to make more changes into the HAL structure and other HAL platforms and I was mainly focused on getting the Teensies up to speed and merged and not every aspect has been abstracted yet. Eventually I'd like to see each of these conditions plugged away.

**Comment by bobc** [comment]
The HAL API will naturally evolve over time as more targets are added, which may mean some reworking  of the existing HAL implementations. Unfortunately that is a "cost" of the HAL approach, but the benefit is we get a clean, maintainable and easily portable application code.

The HAL_Servo should be easy to fix. The interrupt issues need investigation to see what the underlying problems is, whether it is specific to Teensy, K64, Cortex M4, ARM in general, the compiler used etc. When we understand it, then we can create a suitable abstraction.

The problem with "quick fixes" is that they end up getting baked into the code, and then never get resolved.

**Comment by teemuatlut** [comment]
Okey I'll add the few things needed into the HAL structure.

**Comment by teemuatlut** [comment]
I've now added some things into the HAL structure and you won't see any __MK64FX512__ ifdefs in the base code anymore.

I also found a compiling issue that I can't solve. For AVR platform if you enable PINS_DEBUGGING, you're gettings errors that DIO20 and DIO21 are not declared in that scope. And I can't find where these would be, since I didn't find them in the arduino core either. But for Teensies it compiles fine with the debugging on.

**Comment by bobc** [comment]
Looking good! 
I have a Teensy 3.5, when I get a chance I will try it. Do you have any particular hardware (like a shield or adapter) or is it a homebrew setup?

**Comment by teemuatlut** [comment]
There is no hardware available and I'd love to know of a Teensy -> Mega adapter shield.

[This](http://imgur.com/a/n8YLp) was my setup some weeks ago, but I've only changed a few wires since. Much cleaner than the rats nest of jumper wires but still very much DIY. I also had to change the ramps mosfets to get them working on 3.3V.

**Comment by teemuatlut** [comment]
This can be merged if there's nothing further.

**Comment by psavva** [comment]
Do you have a schematic for your setup?

**Comment by teemuatlut** [comment]
I have a drawn schematic that could and should be close to what I have now. You can also look at the fabulous ascii schematic in pins_TEENSY35_36.h.

[Teensy_and_Ramps_pins.zip](https://github.com/MarlinFirmware/Marlin/files/918599/Teensy_and_Ramps_pins.zip)

**Comment by alexxy** [comment]
Hi!

Did you use 3.5 (which is 5v tolerant) or 3.6 version?

**Comment by teemuatlut** [comment]
3.5 exactly because of the tolerance. T3.6 should work as well but I have no way of testing.

**Comment by p3p** [comment]
@thinkyhead I know it's busy times round here stabilising 1.1.x but could these changes be merged, we need to rebase 32bit again before it gets out of hand (and preferably squash ;) )

**Comment by Roxy-3D** [comment]
> I know it's busy times round here stabilising 1.1.x but could these changes be merged, we need to rebase 32bit again before it gets out of hand (and preferably squash ;) )

We have a 'Squash & Merge' button now!   So... that makes it easier on you.   You can keep adding commits until things are 'Right' and they can all be squished together.   I'll do the Squash right now so you can start again with a clean slate.

And you guys have seen this message, right?   https://github.com/MarlinFirmware/Marlin/pull/6373#issuecomment-302824113

**Comment by teemuatlut** [comment]
I know thinkyhead likes to oversee all things Marlin, but would it make sense for someone else to help out and adopt the 32bit effort? Is @bobc still busy with other things?

**Comment by Roxy-3D** [comment]
Well...   We both can help out on the 32-Bit stuff.    And the reason I pointed you to that comment was so you know there is another path available.   The entire 32-Bit code base doesn't have to be working to be merged.   We can make incremental progress towards bringing it up.   

The big, unbreakable rule is...   The 32-Bit stuff **_CAN NOT BREAK_** the 8-bit AVR stuff.   But that should not be hard to accomplish with conditional compilation of the places you hook in.

**Comment by teemuatlut** [comment]
Oh I think you added the link after I had already read your comment.

I actually think the HAL is in a pretty good condition already for a basic setup. All the AVR assembly code is conditional to 8bit and the timers and ADC are handled in the HAL. Mostly what's needed is for some brave souls to start running the 32bit branch. Even if just for 8bit AVR! The thing is that personally I have the TMC2130, TMC2208, Trinamic TRAMS, Teensy3.5, Arduino DUE and I've got an STM32 on the way, and they all are in need for testing and actually running something and I'm here sitting with only one printer.

In the end it shouldn't be too hard to keep the 8bit intact. If something needs to go behind a HAL, you move the existing code to AVR HAL and specialize in your target 32bit HAL.

I guess I'd just like to see the 32bit branch to be in a constant state where all that would be needed is for someone to push the merge button without a complication. It could still be running alongside the main branch but just kept up-to-date. Now this might very well be a bit harder.

**Comment by Roxy-3D** [comment]
> I guess I'd just like to see the 32bit branch to be in a constant state where all that would be needed is for someone to push the merge button without a complication. It could still be running alongside the main branch but just kept up-to-date. Now this might very well be a bit harder.

We can fix that problem easy enough...  The problem with that approach is the ultimate target is to have one code base.   With that said, If you are willing to rebase against the bugfix-v1.1.0 periodically...  I'll merge stuff for you without delay.     If something breaks MarlinFirmware:32-Bit-RCBugFix-new, we can always revert stuff.    We don't need to be as 'protective' of the users of the 32-bit branch as we are for the 8-bit branch.

Mostly, I suggested that other path just because I thought it might help encourage people if they could see some 32-bit stuff working.  Even if it was just some simple stuff like reading configuration settings out of EEPROM (SD-Memory card) and maybe some simple commands.   

But I'm ok with a separate branch too...

I just want to see 32-bit make progress.   I'll help in any way I can.

**Comment by JustAnother1** [comment]
@teemuatlut I would be willing to help out with 32bit. Problem is I don't know in which direction 32bit should move. Not breaking AVR is easy. Not making AVR slower is hard. Also how much functionality may go into the HAL. What should better be outside.

I see several possible ways to do this:

- "a man with a mission" creates a big pull request that implements all of a basic HAL. And everybody else just extends that.

- "a big discussion" is held to find the optimum approach for this.

- "talk is cheap show me the code" a small team of developers agree on an approach and just create one pull request after the other.

All 3 have pro and cons and it seems we lack the people with time and a courage for the first and last option. Then the not merged 32branches could be seen as examples of failed attempts on the first option. Therefore getting 32bit stuff merged is an important signal to the developers.

**Comment by p3p** [comment]
@JustAnother1 The HAL is already integrated with all major components using the C preprocessor so there should be minimal impact on the 8bit codepath, the 32bit platforms just has to implement the necessary api (The Due platform is a guide, though not functional) and is self contained in a folder, there of course is a lot work to be done but I run the 32bit branch and have done for 6 months.

> Even if it was just some simple stuff like reading configuration settings out of EEPROM (SD-Memory card) and maybe some simple commands.

There seems to be some underestimation of how far along the HAL is, possibly because the Due platform doesn't work but that's only because no one with a Due is working on 32bit as far as I know. The hurdle I'm currently working on is building non-Arduino based platforms using platform.io as the STM32 build system is far too complex for "consumers" to be expected to setup and use. Once I have that working and my ReArm arrives I'l get to work on the LPC1768 platform.

**Comment by teemuatlut** [comment]
I'd say

1. Make a huge PR bringing 32bit branch on par with bugfix-1.1.x
2. Agree that
..* AVR needs to work. No exceptions.
..* DUE will be the 'main' 32bit target platform
..* Teensy can tag along for the ride because no hardware for now. Applies to many others too.
3. Keep making smaller and easier PRs keeping the branch up-to-date
4. Make smaller bugfixes to existing HALs.
5. New platforms should ONLY change files in the HAL folder. No preprocessor conditionals targeting a platform unless absolutely must.
6. New HAL functionality as seen fit by the major HAL devs.

Personally I'd like to see a big initial nudge towards main branch and then much smaller and easier to manage updates.
Also I'd like to hear thinkyhead's opinion on his time constraints concerning the 32bit effort. In the end I still don't see it being a democracy as people of the internets have a hard time agreeing on things.
As far as time line goes, I don't see this getting merged into the main branch before the v1.2.0 (?) which should bring us folders. And when we start talking about merging into the main branch, we should be pretty confident on the 8bit HAL already and have some testing done on actual machines. I'd also like to see the overall Marlin branches get clarified with clear release < RC < dev hierarchy.

**Comment by teemuatlut** [comment]
@p3p What's not functional about the DUE? I seem to recall testing some basic stepper movements not too long ago, though I didn't actually do much else.

**Comment by Roxy-3D** [comment]
> Once I have that working and my ReArm arrives I'l get to work on the LPC1768 platform.

YES!!!!

**Comment by JustAnother1** [comment]
@p3p I have seen an commented on bobc's commits. I was then supprised by the 32bit-new branch and still don't understand why we need both. So you are probably right about what works or not.

> The hurdle I'm currently working on is building non-Arduino based platforms using platform.io as the STM32 build system is far too complex for "consumers" to be expected to setup and use. 

I have some STM32 experience and would like to help on that front. Problem is I don't (want to) use Arduino IDE for that. So I#m really just waiting for the (c)make build to work and a visible clue on hot to do HAL in Marlin to get started. 

Once the code is at a point where other developers, like me, can join then that should openly be communicated,...

**Comment by Roxy-3D** [comment]
It would be great if the Arduino IDE was able to load and compile the 32-bit stuff.   But even if that never happens...  Clear direction of how to setup a (c)make environment will be fine.    Please don't let that stop you.   make and makefiles are fine for now.

**Comment by teemuatlut** [comment]
Just curious, why not use the IDE just compiling and uploading? For me the IDE is just a button to see what I've messed up this time but all coding is done on an external editor.

@Roxy-3D The IDE compiles fine for Teensy and DUE and I believe there is even a package for STM32.

One thing that could be cool is a little Marlin branded uploader tool. That's what most basic users use the IDE for anyway.

**Comment by p3p** [comment]
@teemuatlut I don't have that hardware so it was only from what I'd seen on here, I'd happily be wrong on that.

@JustAnother1 The 32bit-new branch is just a rebase, there's nothing really "new" about it other then being more uptodate and although cmake is nice using platform.io as the build system will make it so the end user doesn't need to worry about anything other than hitting compile, it sorts out the framework and libs needed, its the most Arduino-like experience (what marlin users expect) that I can find for 32bit.

**Comment by JustAnother1** [comment]
@Roxy-3D I have not been able to compile Matlin with make or cmake. So I either did not find the correct branch, did not manage to set it up correctly . So I#m not the one to come up with "Clear direction" for it.

@p3p That it is only a rebranch doesn't help at all. The question is why do we need to have both? If there is no reason (as I assume) then please delete one of them so that others do net get confused as I'm already. I'm not too much worried about the end user, but more for the developer. It should be easy for the developer, too.

@teemuatlut  A few years back I looked at Arduino IDE and was stunned how basic it was. After some searching and thinking "maybe I can help them to get syntax highlighting,.. implemented I found a official Arduino statement stating that Arduino IDE was not and will never be directed to professionals. That it does not want those features that make the professional developers productive because they confuse the beginners. I understood that as  direct advice. (I only call my self an professional here because I have been payed to do coding. I don't want to imply that I would be a better coder than other people using Arduino IDE)

So if the decision is to base everything on Arduino IDE and to create "packages" for Arduino IDE for everything else then that might be a valid way forward, just not for me. Decisions like this are the reason for me asking for the direction of the 32bit-HAL.

Again: My way is probably the wrong way. So please do what you think is the best way forward. But I think communicating these decisions / approaches / intentions would help people understand Marlin better and thereby enable them to create better pull-requests and we would end up with less dead end code,...

**Comment by p3p** [comment]
@JustAnother1 platform.io can be used from the command line so can be used with any IDE it is developer friendly, but also has an IDE for consumers to use when they need to build and upload. I've only just started looking into it but so far I'm impressed at how streamlined it can make working with different microcontrollers. I'm just a random guy on the internet I don't get to decide what branches are removed ^^.

**Comment by JustAnother1** [comment]
@p3p I just wanted to make the point that having 2 branches is confusing. I don't care who deletes the branch.

The decision for or against something like platform.io is something that IMHO needs to be done by the "community of developers". I would check in JTAG debug support for all the targets as something that would be more important to me than ease of use. Also performance. The stepper control has rather high requirements,...

**Comment by Roxy-3D** [comment]
The original branch was left in place because there was a belief that some work was in progress against that branch.   But other people needed the code rebased against a more current copy of RCBugFix.  So that was done to help them.    And when some body can further rebase against the currecnt bugfix_v1.1.0 we will probably want that done.    

But "Yes!" we need to clean up unused branches...

**Comment by teemuatlut** [comment]
https://github.com/MarlinFirmware/Marlin/compare/bugfix-1.1.x...teemuatlut:32bit-HAL
I now have a rebased 32bit HAL branch ready that's 'able to merge' into bugfix-1.1.x and differs only with 17 commits in 82 files. It's tested to compile on all three; AVR, Teensy and DUE and it even seems to be working as far as steppers and temperature readings go. Though the DUE was a bit wonky but it'll get sorted for sure.

The thing is, I'm having a hard time getting this 32-Bit-RCBugFix-new branch cleanly updated to my branch. Neither rebase nor merge provided a result I was happy with. So I'd actually suggest that my working branch would supercede the two here on upstream. Any thoughts? I know this would add yet another 32bit branch but I would see it as the cleanest option and I believe it would provide a good base to gradually build upon.

**Comment by Roxy-3D** [comment]
We could change the name of the other two 32-bit branches to have **-deprecated** as a suffix.   At least people would know they are the 'out of favor' branches and if there is no activity for a couple of weeks, we remove them???

@thinkyhead   Your thoughts?   82 files is a lot to look over...  but 32-bit would make much faster progress if it was part of the main branch.    Maybe the thing to do is get a few people to bring it up on 8-Bit AVR machines (Delta, Cartesian and Core ???) and if everything seems to work do the merge????   If we went this route... it might make sense to freeze all other development for a few days while we work through issues until we know everything is OK in 8-bit land.

**Comment by teemuatlut** [comment]
Yes, 82 is a lot. Until you start excluding the actual files that make up the HAL. Take those out, take out files that were just moved or renamed, and you end up with only 15 changed files in the main code!

**Comment by p3p** [comment]
I'm afraid you're removing quite a few defines and abstractions in this rebase (I haven't checked all commits but specifically [Rebase cleanup](https://github.com/teemuatlut/Marlin/commit/5057f327a9b781a943a5b2c6ade0a2cf9edd7048)). the only thing I've struggled with is your modifications to pinsDebug conflicting with thinkyheads, so I've temporarily removed them from my [rebase](https://github.com/p3p/Marlin/tree/32bit-bugfix-1.1.x) I still have a few issues but I think its relatively complete.

**Comment by teemuatlut** [comment]
Yeah the `pinsDebug` has definitely been a lower priority compared to rebase and getting the basic functionality. But I think it shouldn't be too hard to bring that back into working condition.

**Comment by thinkyhead** [comment]
I'm gratified that this is being patched up and can still be made to work. It would be nice to be able to have it as a part of the 1.1.x family so that we can keep it in sync going forward. However… there's this big code reorganization that we really want to do, to break up all the gcodes into separate files and apply a proper file hierarchy. I'm pushing to restrain our further development until we can get the new file layout in place and release it as 1.2.0. And at the same time we can include the HAL as part of that initial 1.2.0 release. The thing is, this is going to take a big sprint to get in place quickly, because we don't want to sideline too many issues for too long. But I think the effort will be very much worth it. I know we're all very anxious to leave the flat file layout behind.

With that in mind, we have a 3-day weekend coming up here in the States…

**Comment by Roxy-3D** [comment]
The big reason I'm supportive of doing a new file hierarchy is because we need a place to put the 32-bit stuff and organize it.   https://github.com/MarlinFirmware/Marlin/pull/6821#issuecomment-303519638     My vote is we freeze development for 2 weeks if it takes that long.   But on the other side of the file hierarchy work, the 32-bit stuff has a place to continue making progress and is part of the main code base.

**Comment by p3p** [comment]
I think it is a good idea to merge the HAL into the main dev branch, stop the constant rebasing and it will help when everyone has to write HAL compatible code, although there may be the down side that major changes will be more difficult and the HAL api is still a mess. I'm very happy to see you are going forward with the restructure, that will help everything. 

My plan was to get the HAL completely independent of Arduino this week after the rebase, but I guess that can wait ^^

**Comment by thinkyhead** [comment]
So… I don't think the initial hierarchy work will take very long, since I've done it once in the past already, and can just follow the same pattern initially. That took a day or two last time, so I expect about the same this time. Once the new hierarchy compiles without issue, that would be the optimum time to add in the HAL. This weekend would be an opportune time to shift over to the 1.2 branch and new hierarchy, as I will have lots of space and time to dive into it.

The main thing that's been holding me back from making any leap is all the issues I feel like I should read and address…
image

**Comment by Roxy-3D** [comment]
Yes...    But I do read just about every issue that shows up in my email Inbox.  There are a few topics I just don't care about and don't read...   My point is there is nothing critical going on right now.  There are no emergencies cooking.   If we freeze all code changes...  There won't be any new emergencies happening.

And people can wait until the reorganization is complete.   

I don't really care how the 32-bit files get organized in the new layout.   But I do feel strongly we should do it as a part of the new organization.   If you really can do the existing stuff in a few days, great!!!  I don't care if we keep the code base frozen for an additional week to get the 32-bit stuff incorporated.   But we need to get the 32-bit stuff limping along and as part of the main branch.

**Comment by fiveangle** [comment]
wasn't @boelle going to try and close some abandoned issues ?

**Comment by p3p** [comment]
@thinkyhead not sure if you want me to put the pull request in, but as I was working on this before the prior discussion.
[This](https://github.com/MarlinFirmware/Marlin/compare/bugfix-1.1.x...p3p:32bit-bugfix-1.1.x) branch brings 32bit up to date with bugfix-1.1.x and I also did a complete squash and split it up into relevant commits to make rebasing easier in this [branch](https://github.com/MarlinFirmware/Marlin/compare/bugfix-1.1.x...p3p:32bit-squash-bugfix-1.1.x), its the same code. 

@teemuatlut it passes the Travis tests (there are some warning that need looked into) but I can't test on hardware here, I also havn't readded your debug pins code as there was considerable refactoring needed.

**Comment by teemuatlut** [comment]
I'm fine with whichever gets used for main branch. You too seem to have the issue that while merging into bugfix-1.1.x would be painless, getting either one of these existing 32bit branches updated would require additional work. Maybe it simple cannot be avoided.

**Comment by p3p** [comment]
These are replacements for the current 32bit branch rather than pull requests because of the necessity of rebasing, keeping the 32bit branch "fastforwardable" from bugfix is more important than maintaining branch history. This is why the 32bit new branch exists @thinkyhead kept it rebased onto bugfix but that always breaks pull requests, usually easily sorted by just cherrypicking the commits from the pull request though. All that's a long winded way of saying I didn't expect my rebased branch to easily merge with the current 32bit branch.

**Pull request merged**
